### PR TITLE
#171: started, to be followed up by @johtoblan

### LINKED ISSUE #171: Require metadata id on the form uri:UUID
Currently, we are not consistently requiring `uri:UUID` formatted metadata ids (e.g., in delete functions). This is confusing, and should be changed.

### PR DESCRIPTION
**Pull request by johtoblan**
**Summary**: Reorders the return of the dmci.app.py::_check_namespace_UUID function
Renames variable metadata_ID to metadata_UUID to be consistent internally

**Related issue**: closes #171 

**Suggested reviewer(s)**:

**Reviewer checklist**:

* [ ] The headers of all files contain a reference to the repository license
* [ ] 100% test coverage of new code - meaning:
  * [ ] The overall test coverage increased or remained the same as before
  * [ ] Every function is accompanied with a test suite
  * [ ] Tests are both positive (testing that the function work as intended with valid data) and negative (testing that the function behaves as expected with invalid data, e.g., that correct exceptions are thrown)
  * [ ] Functions with optional arguments have separate tests for all options
* [ ] Examples are supported by doctests
* [ ] All tests are passing
* [ ] All names (e.g., files, classes, functions, variables) are explicit
* [ ] Documentation (as docstrings) is complete and understandable

The checklist is based on the S-ENDA conventions and definition of done (see [General Conventions](https://s-enda-documentation.readthedocs.io/en/latest/general_conventions.html)). The above points are not necessarily relevant to all contributions. In that case, please add a short explanation to help the reviewer.

### CODE REVIEW

**Comment by johtoblan** [comment]
@mortenwh how should file-distributor name the files? namespace:UUID.xml?

**Comment by mortenwh** [comment]
> @mortenwh how should file-distributor name the files? namespace:UUID.xml?

The filenames are independent from the metadata, right? So I suggest to keep it as it is. I don't see any extra value. Do you?

File: dmci/distributors/distributor.py [hunk 1]
```diff
@@ -117,6 +117,7 @@ def run(self):
     def is_valid(self):
         return self._valid
 
+    # Should be possible to remove this is we consistently use uri:UUID
```
Comment (mortenwh) [inline dmci/distributors/distributor.py:120]: Should this comment be removed? I think my idea was to use the combination of namespace and uuid as identification, but it is maybe also ok like it is done now. In that case the comment can be deleted.
```suggestion
```

**Review by mortenwh — CHANGES_REQUESTED** [review]
Jeg tror det er ok, men fint om vi tar en kort gjennomgang. Ihvertfall en kommentar som kan fjernes antagelig. Hvis jeg forstår det rett, så brukes ikke namespace i distribute, så vi må splitte "<namespace>:<uuid>"?

File: dmci/distributors/distributor.py [hunk 2]
```diff
@@ -99,8 +99,8 @@ def __init__(self, cmd, xml_file=None, metadata_id=None, worker=None, **kwargs):
             self._valid = False
             return
 
-        if self._cmd == DistCmd.DELETE and self._metadata_id is None:
-            logger.error("Command '%s' requires `metadata_id` to be specified" % str(cmd))
+        if self._cmd == DistCmd.DELETE and self._metadata_UUID is None:
```
Comment (mortenwh) [inline dmci/distributors/distributor.py:102]: bør ikke `self._worker._namespace` også sjekkes her?

**Review by mortenwh — CHANGES_REQUESTED** [review]

File: dmci/distributors/distributor.py (hunk 2 above)
Comment (mortenwh) [reply dmci/distributors/distributor.py:102]: må faktisk sjekke at namespace er korrekt også. Nå er det mulig å slette `no.met.dev:*`, selv om namespace er `no.met`

Comment (johtoblan) [reply dmci/distributors/distributor.py:102]: Sjekkes i API

File: dmci/api/app.py [hunk 3]
```diff
@@ -77,8 +77,11 @@ def post_update():
         @self.route("/v1/delete/<metadata_id>", methods=["POST"])
         def post_delete(metadata_id=None):
             """Process delete command."""
-            md_uuid, md_namespace, err = self._check_namespace_UUID(
-                metadata_id)
+            md_namespace, md_uuid, err = self._check_metadata_id(metadata_id,
+                                                                 self._conf.env_string)
+
+            if err is not None:
+                logger.error(err)
```
Comment (johtoblan) [inline dmci/api/app.py:84]: err here is for api-user. logger is output to dmci internally

File: dmci/api/app.py [hunk 4]
```diff
@@ -87,10 +90,7 @@ def post_delete(metadata_id=None):
             else:
                 return self._formatMsgReturn(err), 400
 
-            if err:
-                return self._formatMsgReturn(err), 500
-            else:
-                return self._formatMsgReturn(OK_RETURN), 200
+            return self._formatMsgReturn(OK_RETURN), 200
```
Comment (johtoblan) [inline dmci/api/app.py:93]: Assumes distributor ALWAYS succeeds.
self._distributor_wrapper returns an err to check

File: dmci/api/app.py [hunk 5]
```diff
@@ -206,33 +206,44 @@ def _distributor_wrapper(self, worker):
...
             logger.error(str(e))
+            return None, None, f"Cannot convert to UUID: {uuid_str}"
 
+        # Check that the namespace is correct
         if env_string is not None:
-            if env_string != md_namespace.split(".")[-1]:
-                md_namespace = md_namespace + "." + env_string
+            env = md_namespace.split(".")[-1]
+            if env_string != env:
+                return None, None, f"Dataset metadata_id namespace is wrong: {env}"
```
Comment (johtoblan) [inline dmci/api/app.py:238]: Should or should not the end-user have to know which env he posts to?

File: tests/test_api/test_app.py [hunk 6]
```diff
@@ -255,7 +255,7 @@ def testApiApp_DeleteRequests(client, monkeypatch):
         mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (False, False, [], f, s, e))
 
         response = client.post("/v1/delete/%s" % testUUID, data=MOCK_XML)
-        assert response.status_code == 500
+        assert response.status_code == 500 # this is now 200 and I don't understand why it should be 500...
```
Comment (johtoblan) [inline tests/test_api/test_app.py:258]: data is not None.

File: dmci/api/app.py (hunk 5 above)
Comment (johtoblan) [reply dmci/api/app.py:238]: We do not check and return error in /create /insert /update

File: dmci/api/app.py (hunk 4 above)
Comment (mortenwh) [reply dmci/api/app.py:93]: aha - I see it now..

File: dmci/api/app.py (hunk 3 above)
Comment (mortenwh) [reply dmci/api/app.py:84]: yes - but this is ok, right?

File: dmci/api/app.py (hunk 5 above)
Comment (mortenwh) [reply dmci/api/app.py:238]: no, but delete is different. See the example in the issue (#171). It is very easy to make a mistake, and delete in the wrong environment, and then it becomes all confusing and stressful.. I've done it several times already..|

File: dmci/api/app.py (hunk 3 above)
Comment (johtoblan) [reply dmci/api/app.py:84]: Aren't we already logging something inside of the function? but sure.

Comment (mortenwh) [reply dmci/api/app.py:84]: I removed the logging there to avoid repetition

**Review by mortenwh — APPROVED** [review]
@johtoblan - I think we can merge now.. Do you agree?

**Pull request merged**